File: xphwork/python/run_all.py
# run_all.py - 算法对比主脚本

import time
from config import Config
from load_data import load_data
from random_search import random_search
from genetic_algorithm import genetic_algorithm

def run_all():
    """跑全部算法并输出对比表"""
    print("=" * 50)
    print("  算法对比实验")
    print("=" * 50)
    print()

    # 加载数据
    cfg = Config()
    data = load_data("../data/raw")

    print(f"数据: {cfg.n_drones} 架无人机, {data.n_orders} 件快递")
    print(f"费用: 启用={cfg.enable_cost}, 时间={cfg.alpha}, 超时={cfg.gamma}")
    print()

    methods = [
        ("random_search", "随机×100/架"),
        ("genetic_algorithm", "遗传算法"),
    ]

    results = []
    times = []

    for method_key, method_name in methods:
        print(f"运行 {method_name} ...")
        start = time.time()

        if method_key == "random_search":
            sol = random_search(data, cfg, 100)
        elif method_key == "genetic_algorithm":
            sol = genetic_algorithm(data, cfg)

        elapsed = time.time() - start
        print(f"  完成 ({elapsed:.2f} s)")
        results.append(sol)
        times.append(elapsed)

    # 找最优成本
    costs = [r.total_cost for r in results]
    best_cost = min(costs)

    # 输出对比表
    print()
    print("=" * 70)
    print("  对比结果")
    print("=" * 70)
    print(f"{'方法':<12} {'总成本':>8} {'启用数':>6} {'超时(min)':>8} {'总用时(min)':>10} {'计算耗时(s)':>10}")
    print("-" * 70)

    for m, (method_key, method_name) in enumerate(methods):
        r = results[m]
        gap = (r.total_cost - best_cost) / best_cost * 100 if best_cost > 0 else 0
        print(f"{method_name:<12} {r.total_cost:>8.1f} {r.n_enabled:>6} {r.total_late * 60:>8.1f} {r.total_time * 60:>10.1f} {times[m]:>10.2f}  (gap {gap:.1f}%)")

    print()

    # 输出最优方案的详细配送计划
    best_idx = costs.index(min(costs))
    best = results[best_idx]
    print("=" * 50)
    print(f"  最优方案详细配送计划 ({methods[best_idx][1]})")
    print("=" * 50)

    for k, route in enumerate(best.routes):
        depot_name = route.depot_name if route.depot_name else "驿站"
        print(f"\n无人机 {route.drone} [{depot_name}]  {route.time * 60:.1f}min:")

        if route.details:
            print(f"  详细路径:")
            print(f"  {depot_name}", end="")
            for d in route.details:
                arr_h = int(d["arrive_h"])
                arr_m = round((d["arrive_h"] - arr_h) * 60)
                print(f"\n  -> {d['pickup']}({d['weight']:.1f}kg) -> {d['delivery']} "
                      f"(完成订单 #{d['order_id']:03d}, {arr_h:02d}:{arr_m:02d})", end="")
            print(f"\n  -> {depot_name}")
        else:
            print(f"  路线: {depot_name}", end="")
            for j in route.orders:
                order = data.orders[j - 1]
                print(f" -> {order.pickup_name}({order.weight:.1f}kg) -> {order.delivery_name}", end="")
            print(f" -> {depot_name}")

    print()

if __name__ == "__main__":
    run_all()

File: xphwork/src/algorithms/run_all.m
function run_all()
% RUN_ALL 跑全部算法并输出对比表
%   run_all() 运行 5 种算法并对比结果

    fprintf('============================================\n');
    fprintf('  算法对比实验\n');
    fprintf('============================================\n\n');

    % 添加路径
    src_dir = fileparts(fileparts(mfilename('fullpath')));
    addpath(fullfile(src_dir, 'preprocessing'));
    addpath(fullfile(src_dir, 'utils'));
    addpath(fullfile(src_dir, 'model'));

    % 加载数据
    cfg = config();
    data = load_data();

    fprintf('数据: %d 架无人机, %d 件快递\n', cfg.n_drones, data.n_orders);
    fprintf('费用: 启用=%.1f, 能耗=%.3f, 超时=%.1f, 换电=%.1f\n\n', ...
            cfg.enable_cost, cfg.alpha, cfg.gamma, cfg.swap_cost);

    methods = {'random_search', 'genetic_algorithm'};
    names = {'随机×500/架', '遗传算法'};

    results = cell(length(methods), 1);
    times = zeros(length(methods), 1);

    for m = 1:length(methods)
        fprintf('运行 %s ...\n', names{m});
        tic;
        switch methods{m}
            case 'greedy_assign'
                results{m} = greedy_assign(data, cfg);
            case 'nearest_neighbor'
                results{m} = nearest_neighbor(data, cfg);
            case 'savings'
                results{m} = savings(data, cfg);
            case 'random_search'
                results{m} = random_search(data, cfg, 500);
            case 'genetic_algorithm'
                results{m} = genetic_algorithm(data, cfg);
        end
        times(m) = toc;
        fprintf('  完成 (%.2f s)\n', times(m));
    end

    % 也跑 MILP（如果规模小的话）
    if data.n_orders <= 30
        fprintf('运行 MILP (整数规划) ...\n');
        tic;
        routes = generate_routes(data, cfg);
        [milp_sol, exitflag] = solve_model(routes, data, cfg);
        milp_time = toc;
        if exitflag > 0
            fprintf('  完成 (%.2f s)\n', milp_time);
            has_milp = true;
        else
            fprintf('  无可行解\n');
            has_milp = false;
        end
    else
        has_milp = false;
        fprintf('规模太大，跳过 MILP\n');
    end

    % 输出对比表
    fprintf('\n');
    fprintf('============================================\n');
    fprintf('  对比结果\n');
    fprintf('============================================\n');
    fprintf('%-12s %8s %6s %10s %8s %10s\n', '方法', '总成本', '启用数', '总能耗(Wh)', '超时(min)', '计算耗时(s)');
    fprintf('%-12s %8s %6s %10s %8s %10s\n', '--------', '--------', '------', '----------', '--------', '----------');

    % 找最优成本
    costs = cellfun(@(r) r.total_cost, results);
    if has_milp
        all_costs = [costs, milp_sol.Z];
        best_cost = min(all_costs);
    else
        best_cost = min(costs);
    end

    if has_milp
        gap_milp = 0;
        fprintf('%-12s %8.1f %6d %8.4f %8.2f  (MILP最优)\n', ...
                'MILP', milp_sol.Z, milp_sol.n_enabled, milp_sol.total_late, milp_time);
    end

    for m = 1:length(methods)
        r = results{m};
        if best_cost > 0
            gap = (r.total_cost - best_cost) / best_cost * 100;
            fprintf('%-12s %8.1f %6d %10.0f %8.1f %10.2f  (gap %.1f%%)\n', ...
                    names{m}, r.total_cost, r.n_enabled, r.total_energy, r.total_late*60, times(m), gap);
        else
            fprintf('%-12s %8.1f %6d %10.0f %8.1f %10.2f\n', ...
                    names{m}, r.total_cost, r.n_enabled, r.total_energy, r.total_late*60, times(m));
        end
    end
    fprintf('\n');

    % 输出最优方案的详细配送计划
    [~, best_idx] = min(cellfun(@(r) r.total_cost, results));
    best = results{best_idx};
    fprintf('============================================\n');
    fprintf('  最优方案详细配送计划 (%s)\n', names{best_idx});
    fprintf('============================================\n');

    for k = 1:length(best.routes)
        route = best.routes{k};
        depot_name = '驿站';
        if isfield(route, 'depot_name'), depot_name = route.depot_name; end
        if isfield(route, 'energy')
            fprintf('\n无人机 %d [%s]  %.0fWh:\n', route.drone, depot_name, route.energy);
        else
            fprintf('\n无人机 %d [%s]:\n', route.drone, depot_name);
        end

        if isfield(route, 'details') && ~isempty(route.details)
            fprintf('  详细路径:\n');
            fprintf('  %s', depot_name);
            for kk = 1:length(route.details)
                d = route.details{kk};
                arr_h = floor(d.arrive_h);
                arr_m = round((d.arrive_h - arr_h) * 60);
                fprintf('\n  -> %s(%.1fkg) -> %s (完成订单 #%03d, %02d:%02d)', ...
                        d.pickup, d.weight, d.delivery, d.order_id, arr_h, arr_m);
            end
            fprintf('\n  -> %s\n', depot_name);
        else
            fprintf('  路线: %s', depot_name);
            for kk = 1:length(route.orders)
                j = route.orders(kk);
                fprintf(' -> %s(%.1fkg) -> %s', ...
                        data.orders(j).pickup_name, data.orders(j).weight, ...
                        data.orders(j).delivery_name);
            end
            fprintf(' -> %s\n', depot_name);
        end
    end
    fprintf('\n');
end
